sdk/python/kfp/local/placeholder_utils.py

```diff
@@ -35,17 +35,56 @@ def replace_placeholders(
     """Iterates over each element in the command and replaces placeholders."""
     unique_pipeline_id = make_random_id()
     unique_task_id = make_random_id()
-    return [
-        replace_placeholder_for_element(
+    provided_inputs = get_provided_inputs(executor_input_dict)
+    full_command = [
+        resolve_struct_placeholders(
+            placeholder,
+            provided_inputs,
+        ) for placeholder in full_command
+    ]
+    full_command = flatten_list(full_command)
+    resolved_command = []
+    for el in full_command:
+        resolved_el = resolve_individual_placeholder(
             element=el,
             executor_input_dict=executor_input_dict,
             pipeline_resource_name=pipeline_resource_name,
             task_resource_name=task_resource_name,
             pipeline_root=pipeline_root,
             pipeline_job_id=unique_pipeline_id,
             pipeline_task_id=unique_task_id,
-        ) for el in full_command
-    ]
+        )
+        if resolved_el is None:
+            continue
+        elif isinstance(resolved_el, str):
+            resolved_command.append(resolved_el)
+        elif isinstance(resolved_el, list):
+            resolved_command.extend(resolved_el)
+        else:
+            raise ValueError(
+                f'Got unknown command element {resolved_el} of type {type(resolved_el)}.'
+            )
+    return resolved_command
+
+
+def flatten_list(l: List[Union[str, list, None]]) -> List[str]:
+    """Recursively flattens arbitrarily deeply nested lists, filtering out
```

Comment:
nit: more FYI, ideally avoid recursive implementation in production code cause it could break at some point hitting the recursion depth limit. 
For this specific case, I think it's fine.

Reply:
Changed to an iterative implementation.